### PR TITLE
docs: add fabric nested pipeline example

### LINKED ISSUE #42: Create Fabric pipelines to use for testing
**Goal:** Create Fabric pipelines to use as examples in the repo

**Success criteria:**
[ ] Create Fabric pipelines with typical activities to be used as examples in the repo

### PR DESCRIPTION
**Pull request by ydaponte**
Adding a fabric ingestion pipeline example with a nested pipeline invocation. 
Includes:
- pipelines definitions
- ReadMe 
- pictures of both pipelines
- four tests showcasing activity and pipeline testing with the options: 'should_evaluate_child_pipeline' set to True and to False

Closes #42

### CODE REVIEW

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Nested pipelines
+
+The following example, exemplifies the utilization of nested pipelines including control activities like ForEach and If.
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:3]: Perhaps we can start this page with a short description of what the whole pipeline is trying to achieve. Otherwise the reader has to reverse engineer that from the pipeline and description itself and I am struggling to reverse engineer it myself.

Example: `This pipeline is an example that reads the configuration of a lakehouse, then checks which files are newer or updated and then only downloads those files, uploads them to ADLS and eventually into the Lakehouse.` <- Probably incorrect

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+![Nested Pipelines](pl_main_pipeline.png)
+
+The main pipeline is composed of 2 top activities:
+
+- a lookup activity named "Read Configuration File
+- a ForEach activity names "ForEach"
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:10]:
```suggestion
- a ForEach activity named "ForEach"
```

File: examples/fabric/nested_ingestion_pipeline/fabric/pl_main.DataPipeline/pipeline-content.json [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                    }
+                  ],
+                  "ifFalseActivities": []
+                },
+                "name": "If New Or Updated",
+                "dependsOn": []
+              }
+            ]
+          },
+          "name": "ForEach",
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/pl_main.DataPipeline/pipeline-content.json:99]: Could you have a look at all the names of all activities to make sure they are well defined self-explanatory names? Having the activity names conveying their purpose helps the reader massively.

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to True, a Invoke Pipeline activity named "Invoke Ingestion Pipeline" is triggered.
+
+The Ingestion pipeline is composed also by two activities:
+
+- a Copy activity named "Copy NYCData to ADLS" that copies from the Web to an ADLS Gen2 storage,
+- a Copy activity named "CopyNYCData to Lakehouse" that copies from the ADLS Gen2 storage to the Lakehouse.
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:17]:
```suggestion
- a Copy activity named "Copy NYCData to Lakehouse" that copies from the ADLS Gen2 storage to the Lakehouse.
```

Comment (arjendev) [reply examples/fabric/nested_ingestion_pipeline/README.md:17]: activity already contains the space

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
+# Nested pipelines
+
+The following example, exemplifies the utilization of nested pipelines including control activities like ForEach and If.
+
+![Nested Pipelines](pl_main_pipeline.png)
+
+The main pipeline is composed of 2 top activities:
+
+- a lookup activity named "Read Configuration File
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:9]:
```suggestion
- a Lookup activity named "Read Configuration File
```

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+- a ForEach activity names "ForEach"
+
+Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to True, a Invoke Pipeline activity named "Invoke Ingestion Pipeline" is triggered.
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:12]:
```suggestion
Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to True, an Invoke Pipeline activity named "Invoke Ingestion Pipeline" is triggered.
```

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+- a Copy activity named "CopyNYCData to Lakehouse" that copies from the ADLS Gen2 storage to the Lakehouse.
+
+![Ingestion Pipeline](pl_ingestion_pipeline.png)
+
+## Testing the pl_main activities
+
+The file that contains the individual activity tests for the pl_main_pipeline is located [here](./fabric/tests/test_pl_main_activity.py) and it contains three individual activity tests:
+
+- test_activity_read_configuration_file, which is responsible to test if the json configuration file is correctly located and defined.
+- test_activity_if, responsible to check id the if condition expression is well formed in order to decide wheather or not new or updated data should be ingested.
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:26]:
```suggestion
- test_activity_if, responsible to check if the if condition expression is well formed in order to decide whether new or updated data should be ingested.
```

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+- test_activity_invoke_pipeline, responsible to check if the nested pipeline is correctly invoked.
+
+## Testing the pl_ingestion activities
+
+The file that contains the individual activity tests for the pl_ingestion_pipeline is located [here](./fabric/tests/test_pl_ingestion_activity.py) and it contains two individual activity test:
+
+- test_copy_nyc_data_to_adls2, this test is responsible to test if the source path is correct and id the target path where the data should land in the datalake is well formed and within the correct stablished partition norms.
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:33]:
```suggestion
- test_copy_nyc_data_to_adls2, this test is responsible to test if the source path is correct and if the target path where the data should land in the datalake is well formed and within the correct stablished partition norms.
```

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_ingestion_activity.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    RunParameterType,
+)
+
+
+@pytest.fixture
+def test_framework(request: pytest.FixtureRequest) -> TestFramework:
+    return TestFramework(
+        framework_type=TestFrameworkType.Fabric,
+        # root_folder_path = os.path.join(Path(__file__).parent,"pl_ingestion.DataPipeline")
+        root_folder_path=os.path.join(Path(request.fspath.dirname).parent, "pl_ingestion.DataPipeline"),
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_ingestion_activity.py:19]: Given that we recommend to set the root_folder_path to the root (!) of the data_factory repository, I would suggest to change the root_folder_path to: 

`os.path.join(Path(request.fspath.dirname).parent)`

Same for some other test files, but I will not comment there.

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_activity.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    assert expected_file_name == (activity.type_properties["datasetSettings"]["typeProperties"]["location"]["fileName"])
+
+    expected_folder_name = "config"
+    assert (
+        expected_folder_name
+        == (activity.type_properties["datasetSettings"]["typeProperties"]["location"]["folderPath"])
+    )
+
+
+def test_activity_if(test_framework: TestFramework, pipeline: Pipeline) -> None:
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_activity.py:50]: Should we also introduce a test that tests the negative case? (i.e. result being false)

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_false.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["fileName"]
+        == "lh_config.json"
+    )
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["folderPath"]
+        == "config"
+    )
+
+    # In order to simulate how the activity would look like when the FoarEach activity iterates throught the first row
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_false.py:59]:
```suggestion
    # In order to simulate how the activity would look like when the ForEach activity iterates through the first row
```

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_true.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["fileName"]
+        == "lh_config.json"
+    )
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["folderPath"]
+        == "config"
+    )
+
+    # In order to simulate how the activity would look like when the FoarEach activity iterates throught the first row
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_true.py:59]:
```suggestion
    # In order to simulate how the activity would look like when the ForEach activity iterates through the first row
```

**Review by arjendev — CHANGES_REQUESTED** [review]
Hi @ydaponte, awesome example! I have added some remarks :)

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+- test_copy_nyc_data_to_lakehouse, similar to the previous one but the previous target is now the source and the current target is the Fabric Lakehouse.
+
+## Testing the pl_main pipeline with should_evaluate_child_pipelines=False
+
+When testing a pipeline and instantiating the framework, there is a flag called: should_evaluate_child_pipelines. When setting this flag to False, it means that the test is just being done on the main pipeline and the child pipeline will not be loaded for evaluation. Check the code below:
```
Comment (LeonardHd) [inline examples/fabric/nested_ingestion_pipeline/README.md:38]:
```suggestion
When testing a pipeline and instantiating the framework, there is a flag called: should_evaluate_child_pipelines. When setting this flag to `False`, it means that the test is just being done on the main pipeline and the child pipeline will not be loaded for evaluation. Check the code below:
```

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+```python
+def test_framework(request: pytest.FixtureRequest) -> TestFramework:
+    return TestFramework(
+        framework_type=TestFrameworkType.Fabric,
+        root_folder_path = os.path.join(Path(__file__).parent.parent),
+        should_evaluate_child_pipelines=False
+```
+
+You can find the entire code  [here](./fabric/tests/test_pl_main_pipeline_child_flag_false.py).
```
Comment (LeonardHd) [inline examples/fabric/nested_ingestion_pipeline/README.md:48]:
```suggestion
You can find the entire code [here](./fabric/tests/test_pl_main_pipeline_child_flag_false.py).
```

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+```python
+def test_framework(request: pytest.FixtureRequest) -> TestFramework:
+    return TestFramework(
+        framework_type=TestFrameworkType.Fabric,
+        root_folder_path = os.path.join(Path(__file__).parent.parent),
+        should_evaluate_child_pipelines=True
+```
+
+You can find the entire code  [here](./fabric/tests/test_pl_main_pipeline_child_flag_true.py).
```
Comment (LeonardHd) [inline examples/fabric/nested_ingestion_pipeline/README.md:72]:
```suggestion
You can find the entire code [here](./fabric/tests/test_pl_main_pipeline_child_flag_true.py).
```

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_ingestion_activity.py (hunk 9 above)
Comment (LeonardHd) [reply examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_ingestion_activity.py:19]: nit: Maybe also good to remove the commented out code

**Review by LeonardHd — COMMENTED** [review]
Will leave the approval to @arjendev - looks good to me overall. Just suggested minor changes which should be easy to incorporate by using the GitHub accept suggestions.

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 2 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: Typo corrected and I also I've renamed the ForEach activity to "ForEachYearMonthPair" as per your suggestion on the other comment as well.

File: examples/fabric/nested_ingestion_pipeline/fabric/pl_main.DataPipeline/pipeline-content.json (hunk 3 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/fabric/pl_main.DataPipeline/pipeline-content.json]: Renamed to "ForEachYearMonthPair" to add clarity. I did review the other names and changed to Copy names to be more specific of the source and target on the ingestion pipeline as well.

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 4 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: I've renamed this ones based on the feedback from the previous comment. Now they are called "Copy NYCData from Web to ADLS" and "Copy NYCData from ADLS to Lakehouse" and corrected the typo.

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 5 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: I've also added the closing quote that was missing.

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 6 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: Decided to simplify the sentence. Please check if it is OK now!

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 7 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: Rewrote it to add more clarity. Also added the new case for the if that evaluates to True and for the if that evaluates to False

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 8 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: Rewrote it to add more clarity.

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_ingestion_activity.py (hunk 9 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_ingestion_activity.py]: Good catch! I forgot to change it at the end. I was simplifying the path while testing so it didn't ran on all folders. Done!

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_activity.py (hunk 10 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_activity.py]: Great idea :) - added!

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_false.py (hunk 11 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_false.py]: Corrected as well.

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_true.py (hunk 12 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_true.py]: Corrected

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 13 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: I did some more modifications on this two paragraphs - also added the single quotes on other places where I had True and False referenced. Tks!

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 14 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: Rewrote to "The full example is available here" - sounds better.

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 15 above)
Comment (ydaponte) [reply examples/fabric/nested_ingestion_pipeline/README.md]: Same as previous.

File: examples/fabric/nested_ingestion_pipeline/README.md (hunk 1 above)
Comment (arjendev) [reply examples/fabric/nested_ingestion_pipeline/README.md]: Thanks, that helps!

File: examples/fabric/nested_ingestion_pipeline/README.md [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+If the data is already in the lakehouse and has not been updated, the ingestion pipeline is not triggered.
+
+![Nested Pipelines](pl_main_pipeline.png)
+
+The main pipeline is composed of 2 top activities:
+
+- a Lookup activity named "Read Configuration File"
+- a ForEach activity named "ForEachYearMonthPair"
+
+Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to `True`, the  "Invoke Ingestion Pipeline" activity is triggered.
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/README.md:17]:
```suggestion
Within the ForEach activity, there is a nested If activity called "If New Or Updated" and if this activity evaluates to `True`, the "Invoke Ingestion Pipeline" activity is triggered.
```

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_true.py [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["fileName"]
+        == "lh_config.json"
+    )
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["folderPath"]
+        == "config"
+    )
+
+    # In order to simulate how the activity would look like when the FoarEach activity iterates through the first row
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_true.py:59]:
```suggestion
    # In order to simulate how the activity would look like when the ForEach activity iterates through the first row
```

File: examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_false.py [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["fileName"]
+        == "lh_config.json"
+    )
+    assert (
+        read_configuration_file_activity.type_properties["datasetSettings"]["typeProperties"]["location"]["folderPath"]
+        == "config"
+    )
+
+    # In order to simulate how the activity would look like when the FoarEach activity iterates through the first row
```
Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_false.py:59]:
```suggestion
    # In order to simulate how the activity would look like when the FoarEach activity iterates through the first row
```
```suggestion
    # In order to simulate how the activity would look like when the ForEach activity iterates through the first row
```

Comment (arjendev) [inline examples/fabric/nested_ingestion_pipeline/fabric/tests/test_pl_main_pipeline_child_flag_false.py:59]:
```suggestion
    # In order to simulate how the activity would look like when the ForEach activity iterates through the first row
```

**Review by arjendev — APPROVED** [review]
Thank you so much @ydaponte, let's get this merged :)
